#  Exploratory Data Analysis(EDA)

In [1]:
#Importing necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

C:\Users\tsed\anaconda3\lib\site-packages\scipy\__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.26.4
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


In [2]:
#Loading the dataset 
mpg_df=pd.read_csv('D:\\Tsed\\Python\\Projects\\Mile Per Gallon\\Dataset\\mpg.csv')
mpg_df.head(3)

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite

## 2. Data Cleaning

### 2.1 Handling missing values:  


In [3]:
# detecting null values in horsepower
mpg_df.isna().sum()

mpg             0
cylinders       0
displacement    0
horsepower      6
weight          0
acceleration    0
model_year      0
origin          0
name            0
dtype: int64

In [4]:
# Calculating the threshold (5%) missing value is tolerable to drop them out
threshold=len(mpg_df)*0.05
print('Minimum threshold value:',threshold)
print('Number of missing horsepower values:', mpg_df['horsepower'].isna().sum())

Minimum threshold value: 19.900000000000002
Number of missing horsepower values: 6


In [5]:
# lets print the missing values out
mpg_df[mpg_df['horsepower'].isna()]

mpg  cylinders  displacement  horsepower  weight  acceleration  \
32   25.0          4          98.0         NaN    2046          19.0   
126  21.0          6         200.0         NaN    2875          17.0   
330  40.9          4          85.0         NaN    1835          17.3   
336  23.6          4         140.0         NaN    2905          14.3   
354  34.5          4         100.0         NaN    2320          15.8   
374  23.0          4         151.0         NaN    3035          20.5   

     model_year  origin                  name  
32           71     usa            ford pinto  
126          74     usa         ford maverick  
330          80  europe  renault lecar deluxe  
336          80     usa    ford mustang cobra  
354          81  europe           renault 18i  
374          82     usa        amc concord dl

#### INSIGHT:
    :number of missing value(6) is less than the threshold value(19.9).So we can drop them out

In [6]:
# Dropping the missing values
mpg_df.dropna(inplace=True)
mpg_df.tail(3)

mpg  cylinders  displacement  horsepower  weight  acceleration  \
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model_year origin           name  
395          82    usa  dodge rampage  
396          82    usa    ford ranger  
397          82    usa     chevy s-10

#### INSIGHTS: 
    : The index is still up to 397. Cross checking if the missing values is removed. 
    : If they have been removed, we have to reset the index

In [7]:
# Cross checking if the missing values is removed
mpg_df.isna().any().sum()

0

In [8]:
# Resetting the index to remove empty rows
mpg_df.reset_index(inplace=True)

In [9]:
# Drop index column of the original data 
mpg_df.drop('index',axis=1,inplace=True)

In [10]:
# Cross checking if it the empty rows are removed
mpg_df.tail(3)

mpg  cylinders  displacement  horsepower  weight  acceleration  \
389  32.0          4         135.0        84.0    2295          11.6   
390  28.0          4         120.0        79.0    2625          18.6   
391  31.0          4         119.0        82.0    2720          19.4   

     model_year origin           name  
389          82    usa  dodge rampage  
390          82    usa    ford ranger  
391          82    usa     chevy s-10

#### INSIGHTS: 
     We are good to go!

### 2.2. Formatting & standardizing 

In [11]:
# Standardizing & Converting data types into the proper format
mpg_df['weight']=mpg_df['weight'].astype(float)
mpg_df['model_year']=mpg_df['model_year']+1900
#mpg_df['model_year']=pd.to_datetime(mpg_df['model_year'],errors='coerce').dt.year
print(mpg_df['weight'].dtype)
mpg_df.sample(3)

float64


mpg  cylinders  displacement  horsepower  weight  acceleration  \
110  18.0          3          70.0        90.0  2124.0          13.5   
327  44.6          4          91.0        67.0  1850.0          13.8   
329  29.8          4          89.0        62.0  1845.0          15.3   

     model_year  origin                 name  
110        1973   japan            maxda rx3  
327        1980   japan  honda civic 1500 gl  
329        1980  europe     vokswagen rabbit

### 2.3 Generating new feature from model year

In [12]:
# Generating new feature from model_year column called age
mpg_df.insert(6,'age',datetime.now().year-mpg_df['model_year'])
mpg_df.sample(3)

mpg  cylinders  displacement  horsepower  weight  acceleration  age  \
68   12.0          8         350.0       160.0  4456.0          13.5   52   
207  19.0          4         120.0        88.0  3270.0          21.9   48   
164  13.0          8         302.0       129.0  3169.0          12.0   49   

     model_year  origin                        name  
68         1972     usa  oldsmobile delta 88 royale  
207        1976  europe                 peugeot 504  
164        1975     usa             ford mustang ii

### 2.4 Handling Outliers

In [25]:
upper_threshold={}
lower_threshold={}

for i in mpg_df.select_dtypes('number').columns:
    upper_threshold[i]=(mpg_df[i].quantile(0.75)+(1.5*(mpg_df[i].quantile(0.75)-mpg_df[i].quantile(0.25))))
    lower_threshold[i]=(mpg_df[i].quantile(0.25)-(1.5*(mpg_df[i].quantile(0.75)-mpg_df[i].quantile(0.25))))

outliers_dfs=outliers_df.append(upper_threshold,ignore_index=True,)
outliers_dfss=outliers_dfs.append(lower_threshold,ignore_index=True)

outliers_dfss.index=['min','Q1','Q2','Q3','max','upper_threshold','lower_threshold']

print('INSGIHTS: The data has taken for ',mpg_df['model_year'].nunique(),' model years')
print('From',mpg_df['model_year'].min(),'To',mpg_df['model_year'].max())
outliers_dfss.iloc[:,:]

INSGIHTS: The data has taken for  13  model years
From 1970 To 1982


mpg  cylinders  displacement  horsepower   weight  \
min               9.00        3.0        68.000        46.0  1613.00   
Q1               17.00        4.0       105.000        75.0  2225.25   
Q2               22.75        4.0       151.000        93.5  2803.50   
Q3               29.00        8.0       275.750       126.0  3614.75   
max              46.60        8.0       455.000       230.0  5140.00   
upper_threshold  47.00       14.0       531.875       202.5  5699.00   
lower_threshold  -1.00       -2.0      -151.125        -1.5   141.00   

                 acceleration   age  model_year  
min                     8.000  42.0      1970.0  
Q1                     13.775  45.0      1973.0  
Q2                     15.500  48.0      1976.0  
Q3                     17.025  51.0      1979.0  
max                    24.800  54.0      1982.0  
upper_threshold        21.900  60.0      1988.0  
lower_threshold         8.900  36.0      1964.0

INSIGHTS: 

          :For categorical variable outlier theshold values shouldnt be used eg: cylinders column the thresholds doesnt make 

          sense there couldnt be vehicles with -2 or 14 cylinders.

          : The upper threshold value of horsepower is lower than the max value of hp so there is outlier
          
          : There is outliers in acceleration explanatory variable in both direction

In [34]:
print('number of outliers above threshold:',len(mpg_df[mpg_df['horsepower']>202.5]))
mpg_df[mpg_df['horsepower']>202.5]

number of outliers above threshold: 10


mpg  cylinders  displacement  horsepower  weight  acceleration  age  \
6    14.0          8         454.0       220.0  4354.0           9.0   54   
7    14.0          8         440.0       215.0  4312.0           8.5   54   
8    14.0          8         455.0       225.0  4425.0          10.0   54   
13   14.0          8         455.0       225.0  3086.0          10.0   54   
25   10.0          8         360.0       215.0  4615.0          14.0   54   
27   11.0          8         318.0       210.0  4382.0          13.5   54   
66   11.0          8         429.0       208.0  4633.0          11.0   52   
93   13.0          8         440.0       215.0  4735.0          11.0   51   
94   12.0          8         455.0       225.0  4951.0          11.0   51   
115  16.0          8         400.0       230.0  4278.0           9.5   51   

     model_year origin                          name  
6          1970    usa              chevrolet impala  
7          1970    usa             plymouth fury iii  
8          1970    usa              pontiac catalina  
13         1970    usa       buick estate wagon (sw)  
25         1970    usa                     ford f250  
27         1970    usa                    dodge d200  
66         1972    usa               mercury marquis  
93         1973    usa  chrysler new yorker brougham  
94         1973    usa      buick electra 225 custom  
115        1973    usa            pontiac grand prix

In [62]:
#mpg_df[mpg_df.duplicated('name')]['name'].value_counts()

In [73]:
#print('Number of outliers above& below threshold:',len((mpg_df[(mpg_df['acceleration']>21) | (mpg_df['acceleration']<8.9)])))
print('Total number of outliers:',len(mpg_df[((mpg_df['acceleration']>21) | (mpg_df['acceleration']<8.9) )|(mpg_df['horsepower']>202.5)]))
outliers=mpg_df[((mpg_df['acceleration']>21) | (mpg_df['acceleration']<8.9) )|(mpg_df['horsepower']>202.5)]
outliers

Total number of outliers: 23


mpg  cylinders  displacement  horsepower  weight  acceleration  age  \
6    14.0          8         454.0       220.0  4354.0           9.0   54   
7    14.0          8         440.0       215.0  4312.0           8.5   54   
8    14.0          8         455.0       225.0  4425.0          10.0   54   
9    15.0          8         390.0       190.0  3850.0           8.5   54   
11   14.0          8         340.0       160.0  3609.0           8.0   54   
13   14.0          8         455.0       225.0  3086.0          10.0   54   
25   10.0          8         360.0       215.0  4615.0          14.0   54   
27   11.0          8         318.0       210.0  4382.0          13.5   54   
58   23.0          4          97.0        54.0  2254.0          23.5   52   
66   11.0          8         429.0       208.0  4633.0          11.0   52   
93   13.0          8         440.0       215.0  4735.0          11.0   51   
94   12.0          8         455.0       225.0  4951.0          11.0   51   
115  16.0          8         400.0       230.0  4278.0           9.5   51   
193  29.0          4          85.0        52.0  2035.0          22.2   48   
194  24.5          4          98.0        60.0  2164.0          22.1   48   
207  19.0          4         120.0        88.0  3270.0          21.9   48   
242  43.1          4          90.0        48.0  1985.0          21.5   46   
297  27.2          4         141.0        71.0  3190.0          24.8   45   
298  23.9          8         260.0        90.0  3420.0          22.2   45   
323  44.3          4          90.0        48.0  2085.0          21.7   44   
324  43.4          4          90.0        48.0  2335.0          23.7   44   
326  30.0          4         146.0        67.0  3250.0          21.8   44   
388  44.0          4          97.0        52.0  2130.0          24.6   42   

     model_year  origin                               name  
6          1970     usa                   chevrolet impala  
7          1970     usa                  plymouth fury iii  
8          1970     usa                   pontiac catalina  
9          1970     usa                 amc ambassador dpl  
11         1970     usa                 plymouth 'cuda 340  
13         1970     usa            buick estate wagon (sw)  
25         1970     usa                          ford f250  
27         1970     usa                         dodge d200  
58         1972  europe                  volkswagen type 3  
66         1972     usa                    mercury marquis  
93         1973     usa       chrysler new yorker brougham  
94         1973     usa           buick electra 225 custom  
115        1973     usa                 pontiac grand prix  
193        1976     usa                 chevrolet chevette  
194        1976     usa                    chevrolet woody  
207        1976  europe                        peugeot 504  
242        1978  europe    volkswagen rabbit custom diesel  
297        1979  europe                        peugeot 504  
298        1979     usa  oldsmobile cutlass salon brougham  
323        1980  europe               vw rabbit c (diesel)  
324        1980  europe                 vw dasher (diesel)  
326        1980  europe                 mercedes-benz 240d  
388        1982  europe                          vw pickup

Let's remove these outliers

In [77]:
mpg_df_cleaned=mpg_df.drop(outliers.index)
mpg_df_cleaned

mpg  cylinders  displacement  horsepower  weight  acceleration  age  \
0    18.0          8         307.0       130.0  3504.0          12.0   54   
1    15.0          8         350.0       165.0  3693.0          11.5   54   
2    18.0          8         318.0       150.0  3436.0          11.0   54   
3    16.0          8         304.0       150.0  3433.0          12.0   54   
4    17.0          8         302.0       140.0  3449.0          10.5   54   
..    ...        ...           ...         ...     ...           ...  ...   
386  27.0          4         151.0        90.0  2950.0          17.3   42   
387  27.0          4         140.0        86.0  2790.0          15.6   42   
389  32.0          4         135.0        84.0  2295.0          11.6   42   
390  28.0          4         120.0        79.0  2625.0          18.6   42   
391  31.0          4         119.0        82.0  2720.0          19.4   42   

     model_year origin                       name  
0          1970    usa  chevrolet chevelle malibu  
1          1970    usa          buick skylark 320  
2          1970    usa         plymouth satellite  
3          1970    usa              amc rebel sst  
4          1970    usa                ford torino  
..          ...    ...                        ...  
386        1982    usa           chevrolet camaro  
387        1982    usa            ford mustang gl  
389        1982    usa              dodge rampage  
390        1982    usa                ford ranger  
391        1982    usa                 chevy s-10  

[369 rows x 10 columns]

## 3.Visualization